File: Binary Search Tree/binary_search_tree.rb
require_relative 'bst_node'

class BinarySearchTree
  attr_reader :root

  def initialize
    @root = nil
  end

  def insert(value)
    node = BSTNode.new(value)
    @root ? @root.append(node) : @root = node
    balance()
  end

  def delete(value)
    parent, target = search(value)

    if num_children(@root) == 0
      @root = nil

    elsif num_children(target) == 0
      parent.left = nil if target.value < parent.value
      parent.right = nil if target.value > parent.value
      
    elsif num_children(target) == 1
      parent.left = target.left || target.right if target.value < parent.value
      parent.right = target.left || target.right if target.value > parent.value

    elsif num_children(target) == 2
      proxy = maximum(target.left)
      delete(proxy.value)
      target.value < parent.value ? parent.left = proxy : parent.right = proxy
      proxy.left = target.left
      proxy.right = target.right
    end
    
    balance()
  end

  def find(value, node = @root)
    return nil unless node
    return node if node.value == value
    return value < node.value ? find(value, node.left) : find(value, node.right)
  end

  def balance
  end

  def maximum(node = @root)
    while node.right
      node = node.right
    end
    node
  end

  def depth(node = @root)
    return -1 unless node
    left = depth(node.left)
    rite = depth(node.right)
    return left >= rite ? left + 1 : rite + 1
  end 

  def is_balanced?(node = @root)
    return true unless node
    left = depth(node.left)
    rite = depth(node.right)
    is_depth = (left - rite).abs <= 1
    is_balan = is_balanced?(node.left) && is_balanced?(node.right)
    return is_depth && is_balan
  end

  def in_order_traversal(node = @root, arr = [])
    return arr unless node
    in_order_traversal(node.left, arr)
    arr.push(node.value)
    in_order_traversal(node.right, arr)
  end

  def pre_order_traversal(node = @root, arr = [])
    return arr unless node
    in_order_traversal(node.left, arr)
    in_order_traversal(node.right, arr)
    arr.push(node.value)
  end

  def post_order_traversal(node = @root, arr = [])
    return arr unless node
    arr.push(node.value)
    in_order_traversal(node.left, arr)
    in_order_traversal(node.right, arr)
  end

  private
  def search(value)
    parent = nil
    target = @root
    while target
      break if target.value == value
      parent = target
      target = value < target.value ? target.left : target.right
    end
    [parent, target]
  end

  def num_children(node)
    2 - [node.left, node.right].count(nil)
  end
end

File: Binary Search Tree/BinarySearchTree.cpp
#include "BinarySearchTree.hpp"

/* constructor */
BinarySearchTree::BinarySearchTree(BSTNode* root) :
  root{root} {
}

BinarySearchTree::~BinarySearchTree() {
  if (root) delete root;
}

/* basic operations */
// O(log(n))
BSTNode* BinarySearchTree::insert(int value) {
  BSTNode* node = new BSTNode(value);
  if (root == nullptr) {
    root = node;
  } else {
    root->append(node);
  }
  return node;
}

// O(log(n))
void BinarySearchTree::remove(int value) {
  std::vector<BSTNode*> pair = getTarget(value);
  BSTNode* target = pair[0];
  BSTNode* parent = pair[1];
  
  if (target == root) {
    BSTNode* replacement = root->left->max();
    remove(replacement->value);
    replacement->left = root->left;
    replacement->rite = root->rite;
    root = replacement;
    return;
  }
  
  bool isLeft = false;
  if (target->value <= parent->value) isLeft = true;
  
  if (target->children().size() == 0) {
    isLeft ? parent->left = nullptr : parent->rite = nullptr;
    return;
  }
  
  if (target->children().size() == 1) {
    if (target->left) {
      isLeft ? parent->left = target->left : parent->rite = target->left;
    } else {
      isLeft ? parent->left = target->rite : parent->rite = target->rite;
    }
    return;
  }
  
  if (target->children().size() == 2) {
    BSTNode* replacement = target->left->max();
    remove(replacement->value);
    replacement->left = target->left;
    replacement->rite = target->rite;
    isLeft ? parent->left = replacement : parent->rite = replacement;
  }
}

// O(log(n))
BSTNode* BinarySearchTree::find(int value) const {
  BSTNode* node = root;
  while (node) {
    if (node->value == value) return node;
    node = (value < node->value) ? node->left : node->rite;
  }
  return nullptr;
}

// O(log(n))
bool BinarySearchTree::includes(int value) const {
  BSTNode* node = root;
  while (node) {
    if (node->value == value) return true;
    node = (value < node->value) ? node->left : node->rite;
  }
  return false;
}

std::vector<BSTNode*> BinarySearchTree::getTarget(int value) const {
  BSTNode* target = root;
  BSTNode* parent = nullptr;
  std::vector<BSTNode*> pair;
  
  while (target) {
    if (target->value == value) {
      pair.push_back(target);
      pair.push_back(parent);
      return pair;
    }
    parent = target;
    target = (value < target->value) ? target->left : target->rite;
  }
  
  return pair;
}

int BinarySearchTree::depth(BSTNode* node) const {
  if (!node) return 0;
  int left = depth(node->left) + 1;
  int rite = depth(node->rite) + 1;
  return left >= rite ? left : rite;
}

bool BinarySearchTree::isBalanced(BSTNode* node) const {
  if (!node) return true;
  int left = depth(node->left);
  int rite = depth(node->rite);
  int difference = abs(left - rite);
  
  if (difference <= 1) {
    bool leftBal = isBalanced(node->rite);
    bool riteBal = isBalanced(node->left);
    return leftBal && riteBal;
  } else {
    return false;
  }
}
